src/main/java/com/iota/iri/service/garbagecollector/GarbageCollector.java

```diff
@@ -0,0 +1,323 @@
+     * @param queueProcessor function that takes care of processing the queue for this particular type
+     */
+    void registerQueueProcessor(Class<? extends GarbageCollectorJob> jobClass, QueueProcessor queueProcessor) {
+        this.queueProcessors.put(jobClass, queueProcessor);
+    }
+
+    /**
+     * This method allows us to register a {@link QueueConsolidator} for the given job type.
+     *
+     * Some jobs can be consolidated to consume less space in the queue by grouping them together or skipping them
+     * completely. While the consolidation of multiple jobs into fewer ones is optional and only required for certain
+     * types of jobs, this method allows us to generically handle this use case by registering a handler for the job
+     * class that supports this feature.
+     *
+     * @param jobClass class of the job that the GarbageCollector shall be able to handle
+     * @param queueConsolidator lambda that takes care of consolidating the entries in a queue
+     */
+    void registerQueueConsolidator(Class<? extends GarbageCollectorJob> jobClass, QueueConsolidator queueConsolidator) {
+        this.queueConsolidators.put(jobClass, queueConsolidator);
+    }
+
+    /**
+     * This method persists the changes of the garbage collector so IRI can continue cleaning up upon restarts.
+     *
+     * Since cleaning up the old database entries can take a long time, we need to make sure that it is possible to
+     * continue where we stopped. This method therefore writes the state of the queued jobs into a file that is read
+     * upon re-initialization of the GarbageCollector.
+     *
+     * @throws GarbageCollectorException if something goes wrong while writing the state file
+     */
+    void persistChanges() throws GarbageCollectorException {
+        try {
+            Files.write(
+                Paths.get(getStateFile().getAbsolutePath()),
+                () -> garbageCollectorJobs.entrySet().stream().flatMap(
+                    entry -> entry.getValue().stream()
+                ).<CharSequence>map(
+                    entry -> entry.getClass().getCanonicalName() + ";" + entry.serialize()
+                ).iterator()
+            );
+        } catch(IOException e) {
+            throw new GarbageCollectorException("could not persists garbage collector state", e);
+        }
+    }
+
+    /**
+     * This method contains the logic for scheduling the jobs and executing them.
+     *
+     * It iterates through all available queues and executes the corresponding {@link QueueProcessor} for each of them
+     * until all queues have been processed.
+     *
+     * @throws GarbageCollectorException if anything goes wrong while processing the cleanup jobs
+     */
+    private void processCleanupJobs() throws GarbageCollectorException {
+        for(Map.Entry<Class<? extends GarbageCollectorJob>, ArrayDeque<GarbageCollectorJob>> entry : garbageCollectorJobs.entrySet()) {
```

Comment:
ArrayDeque = Queue

Reply:
Since i need things like addLast and so on I kind of need to be specific here.